src/blocks/packages.rs

```diff
@@ -0,0 +1,356 @@
+    pub package_manager: Vec<PackageManager>,
+    pub format: FormatConfig,
+    pub format_singular: FormatConfig,
+    pub format_up_to_date: FormatConfig,
+    pub warning_updates_regex: Option<String>,
+    pub critical_updates_regex: Option<String>,
+    pub ignore_updates_regex: Option<String>,
+    pub ignore_phased_updates: bool,
+    pub aur_command: Option<String>,
+}
+
+#[derive(Deserialize, Debug, Clone, Copy, PartialEq)]
+#[serde(rename_all = "lowercase")]
+pub enum PackageManager {
+    Apt,
+    Pacman,
+    Aur,
+}
+
+pub async fn run(config: &Config, api: &CommonApi) -> Result<()> {
+    let mut config: Config = config.clone();
+
+    let format = config.format.with_default(" $icon $total.eng(w:1) ")?;
+    let format_singular = config
+        .format_singular
+        .with_default(" $icon $total.eng(w:1) ")?;
+    let format_up_to_date = config
+        .format_up_to_date
+        .with_default(" $icon $total.eng(w:1) ")?;
+
+    // If user provide package manager in any of the formats then consider that also
+    macro_rules! any_format_contains {
+        ($name:expr) => {
+            format.contains_key($name)
+                || format_singular.contains_key($name)
+                || format_up_to_date.contains_key($name)
+        };
+    }
+
+    let apt = any_format_contains!("apt");
+    let aur = any_format_contains!("aur");
+    let pacman = any_format_contains!("pacman");
+
+    if !config.package_manager.contains(&PackageManager::Apt) && apt {
+        config.package_manager.push(PackageManager::Apt);
+    }
+    if !config.package_manager.contains(&PackageManager::Pacman) && pacman {
+        config.package_manager.push(PackageManager::Pacman);
+    }
+    if !config.package_manager.contains(&PackageManager::Aur) && aur {
+        config.package_manager.push(PackageManager::Aur);
+    }
+
+    let warning_updates_regex = config
+        .warning_updates_regex
+        .as_deref()
+        .map(Regex::new)
+        .transpose()
+        .error("invalid warning updates regex")?;
+    let critical_updates_regex = config
+        .critical_updates_regex
+        .as_deref()
+        .map(Regex::new)
+        .transpose()
+        .error("invalid critical updates regex")?;
+    let ignore_updates_regex = config
+        .ignore_updates_regex
+        .as_deref()
+        .map(Regex::new)
+        .transpose()
+        .error("invalid ignore updates regex")?;
+
+    let mut package_manager_vec = Vec::new();
+
+    for &package_manager in config.package_manager.iter() {
+        let backend = match package_manager {
+            PackageManager::Apt => Box::new(Apt::new().await?) as Box<dyn Backend>,
+            PackageManager::Pacman => Box::new(Pacman::new().await?) as Box<dyn Backend>,
+            PackageManager::Aur => Box::new(Aur::new()) as Box<dyn Backend>,
+        };
+
+        package_manager_vec.push(backend);
+    }
+
+    loop {
+        let (mut apt_count, mut pacman_count, mut aur_count) = (0, 0, 0);
```

Comment:
I think it's better to have a single place with driver-specific code in the block and that is when creating the drivers. Here you can have a `HashMap<&'static str, u32>`, so that then you don't have to `match` on the exact type of each driver. But `Backend::package_manager()` (which can be shortened to just `Backend::name()`) will have to return `&'static str`.

Reply:
How about if we have HashMap of (backend,update_count) at the starting here
```rust
    let mut package_manager_vec = Vec::new();

    for &package_manager in config.package_manager.iter() {
        let backend = match package_manager {
            PackageManager::Apt => Box::new(Apt::new().await?) as Box<dyn Backend>,
            PackageManager::Pacman => Box::new(Pacman::new().await?) as Box<dyn Backend>,
            PackageManager::Aur => Box::new(Aur::new()) as Box<dyn Backend>,
        };

        package_manager_vec.push(backend);
    }
```
so that we have hashmap which contains what we need till the end. and we can get the name of backend by `Backend::package_manager()`

Reply:
What I hand in mind is something like

```rust
for package_manager in &package_manager_vec {
    let updates = package_manager.get_updates_list().await?;
    values.insert(
        package_manager.name(),
        Value::number(package_manager.get_update_count(&updates).await?),
    );
    warning |= warning_updates_regex
        .as_ref()
        .is_some_and(|regex| package_manager.has_matching_update(&updates, regex));
    critical |= critical_updates_regex
        .as_ref()
        .is_some_and(|regex| package_manager.has_matching_update(&updates, regex));
}
```

Reply:
Done